pkg/macaroon/macaroon.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (C) 2019 Storj Labs, Inc. // See LICENSE for copying information.
+
+package macaroon
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/subtle"
+)
+
+// Macaroon is a struct that determine contextual caveats and authorization
+type Macaroon struct {
+	head    []byte
+	caveats []Caveat
+	tail    []byte
+}
+
+// Caveat is a struct that determine restriction
+type Caveat struct {
```

Comment:
so, you'll want a Caveat to have lots of things in it, potentially being defined as a protobuf. something like https://github.com/storj/storj/blob/0dc098cd652bb5e80f025f2755026c419a3a77ea/lib/uplink/apikey.go#L33

so then the caveat will need a serialization mechanism so that you can turn it into a []byte for signing.

Reply:
But doesn't caveat should contain only one restriction in order for the process of creating more restricting macaroons to work correctly? Also, the restriction part you are talking about is what ID of the caveat is, there is also VerificationID and optional Location. Those parts are described in the official documentation. VerificationID is needed when you have a third party macaroon, for example. Anyway on this layer we don't care what restrictions caveat contains in it's ID because it should not affect the general flow of authorizing and serializing of the macaroon.

Reply:
> But doesn't caveat should contain only one restriction in order for the process of creating more restricting macaroons to work correctly? 
a caveat does not need to only have one restriction, no

> Anyway on this layer we don't care what restrictions caveat contains in it's ID because it should not affect the general flow of authorizing and serializing of the macaroon.
okay, but then why have a caveat struct?